gnomad_qc/v4/create_release/variant_statistics.py

```diff
@@ -0,0 +1,286 @@
+
+    :param test: Whether to use a tmp path for analysis of the test VDS instead of the full v4 VDS
+    :param data_type: Data type used in sample QC, e.g. "exomes" or "genomes"
+    :return: Per sample variant counts table
+    """
+    return VersionedTableResource(
+        CURRENT_RELEASE,
+        {
+            version: TableResource(
+                f"{_release_root(version=version, test=test, data_type=data_type, extension='ht')}/gnomad.{data_type}.v{version}.per_sample_variant_counts.ht"
+            )
+            for version in RELEASES
+        },
+    )
+
+
+def per_sample_datatype(
+    data_type: str = "exomes",
+    test: bool = False,
+    overwrite: bool = False,
+    filter_variant_qc: bool = False,
+    filter_calling_intervals: bool = False,
+    singletons_by_pop: bool = False,
+) -> None:
+    """
+    Write out Hail's sample_qc output for desired samples and variants.
+    Useful for finding the amount of variants (total or specific types) per sample.
+    Also prints distribution of total variants called per sample.
+
+    :param data_type: String of either "exomes" or "genomes" for the sample type.
+    :param test: Boolean for if you would like to use a small test set.
+    :param overwrite: Boolean to overwrite checkpoint files if requested.
+    :param filter_variant_qc: Bool to only count variants that pass all variant qc filters (i.e. VQSR, AC0).
+    :param filter_calling_intervals: Bool to filter only to variants in calling intervals.
+    :param singletons_by_pop: Bool to calculate the mean number of singletons by population group.
+    """
+
+    # Read in release data and metadata and VDS
+    ht_release_data = release_sites(data_type).ht()
+    if test:
+        logger.info("Test: filtering to only chr22 sites and 467 test samples in VDS")
+        ht_release_data = ht_release_data.filter(
+            ht_release_data.locus.contig == "chr22"
+        )
+
+    if filter_variant_qc:
+        logger.info("Filter to those which pass all filters")
+        ht_release_data = ht_release_data.filter(hl.len(ht_release_data.filters) == 0)
+
+    if filter_calling_intervals:
+        logger.info("Filter to calling intervals")
+        calling_interval_ht = calling_intervals(
+            interval_name="union",
+            calling_interval_padding=0,
+        ).ht()
+        ht_release_data = ht_release_data.filter(
+            ~hl.is_missing(calling_interval_ht[ht_release_data.locus])
+        )
+
+    meta_ht = meta(data_type=data_type).ht()
+    vds_data = get_gnomad_v4_vds(test=test, release_only=True)
+    vds_data_mt = vds_data.variant_data
+
+    # Filter to only variants and samples in release data
+    vds_data_filtered = vds_data_mt.filter_rows(
+        ~hl.is_missing(ht_release_data[vds_data_mt.row_key])
+    )
+
+    vds_data_filtered = vds_data_filtered.filter_cols(
+        meta_ht[vds_data_filtered.s].release
+    )
+
+    vds_data_filtered = vds_data_filtered.annotate_cols(
```

Comment:
With the suggestion below for splitting this function up, this annotation can just be added to the HT in the aggregation function since it isn't needed for computing the per sample stats